### ISSUE TITLE
Querying Gurobi attributes returns OptimizeNotCalled() error 

### ISSUE THREAD

**Issue by rezabayani** [jump-dev/MathOptInterface.jl#2361]
I tried running an older code (first written around 2 years ago) with newer JuMP and Gurobi versions. When I try to access some of the model attributes such as Gurobi.VariableAttribute("LB") or Gurobi.ConstraintAttribute("RHS"), I get the error OptimizeNotCalled(). This worked fine previously. 

```julia
model = direct_model(Gurobi.Optimizer())
@variable(model, x >= 1.5)
@constraint(model, c, 2x + 1 <= 10)
@objective(model, Min, 1-x)

get_optimizer_attribute(model, Gurobi.ModelAttribute("IsMIP")) # ERROR: OptimizeNotCalled()
MOI.get.(model, Gurobi.VariableAttribute("LB"), x) # ERROR: OptimizeNotCalled()
MOI.get.(model, Gurobi.ConstraintAttribute("RHS"), c) # ERROR: OptimizeNotCalled()
```
Right now, I need to run optimize! and then it works as intended. Is there a workaround for this issue? 

My current installed versions are: 
JuMP version = "1.16.0"
Gurobi version = "1.2.0"

My old code versions are: 
JuMP version = "0.7.3"
Gurobi version = "0.10.3"

**Comment by odow** [comment]
So you can force this with:
```julia
julia> using JuMP, Gurobi

julia> model = direct_model(Gurobi.Optimizer())
A JuMP Model
Feasibility problem with:
Variables: 0
Model mode: DIRECT
Solver name: Gurobi

julia> @variable(model, x >= 1.5)
x

julia> @constraint(model, c, 2x + 1 <= 10)
c : 2 x ≤ 9

julia> @objective(model, Min, 1-x)
-x + 1

julia> MOI.get(backend(model), Gurobi.ModelAttribute("IsMIP"))
0

julia> MOI.get.(backend(model), Gurobi.VariableAttribute("LB"), index.(x))
1.5

julia> MOI.get.(backend(model), Gurobi.ConstraintAttribute("RHS"), index.(c))
9.0
```

I wonder if JuMP needs a `get_attribute(model, args...; ignore_optimize_not_called = true)` method.

**Comment by odow** [comment]
The change in Gurobi was part of v1.0.0, so it was part of a breaking release and so technically okay. I've thought of a few options but nothing seems like an obviously good idea.

Maybe we just need to update https://github.com/jump-dev/Gurobi.jl#accessing-gurobi-specific-attributes

**Comment by rezabayani** [comment]
> So you can force this with:
> julia> MOI.get(backend(model), Gurobi.ModelAttribute("IsMIP"))
> 0
> julia> MOI.get.(backend(model), Gurobi.VariableAttribute("LB"), index.(x))
> 1.5
> julia> MOI.get.(backend(model), Gurobi.ConstraintAttribute("RHS"), index.(c))
> 9.0

Thanks, Oscar, this is what I was looking for.

**Comment by odow** [comment]
I've updated the Gurobi docs: https://github.com/jump-dev/Gurobi.jl/pull/539. Not sure what we should do here.

**Comment by blegat** [comment]
> We don't have a way to say which attributes were set by the solver and require `optimize!`, and which are defined by the optimizer without needing `optimize!`.

Can you elaborate? Can't we say that the ones of this issue don't need optimize ?

**Comment by odow** [comment]
We know. But MOI doesn't have a `MOI.is_set_by_optimizer_but_doesnt_need_optimize` function.

**Comment by blegat** [comment]
I see, maybe `is_set_by_optimize` is really "CachingOptimizer` should forward it to the `optimizer` and we shouldn't have done the JuMP's check that prevents getting it when the status is `OPTIMIZE_NOT_CALLED`. The right approach for that was to have a custom error defined in MOI that is thrown by MOI wrappers when appropriate.

**Comment by odow** [comment]
Agreed we could change for MOI 2.0. Anything non-breaking in the near term?

**Comment by blegat** [comment]
> Agreed we could change for MOI 2.0

Does that need any change at the MOI level ? Also, it doesn't seem breaking to me. Only current JuMP code that error will instead work.

**Comment by odow** [comment]
I was thinking it would be breaking to change from `is_set_by_optimize` to the new method.

**Comment by blegat** [comment]
I would be the same method. Renaming it can be in MOI 2.0

**Comment by odow** [comment]
I don't really understand what you're suggesting. That JuMP should just stop throwing the error and forward to the solver? But then it might start returning incorrect results.

**Comment by blegat** [comment]
> That JuMP should just stop throwing the error and forward to the solver? But then it might start returning incorrect results.

Yes but that's not breaking because it's already the case if the MOI wrapper is used without JuMP. This JuMP errors just makes existing bugs be more rare (because they don't occur when the solver is used from JuMP) instead of fixing them.

**Comment by odow** [comment]
> Yes but that's not breaking because it's already the case if the MOI wrapper is used without JuMP

It is, because we decided that at the JuMP level we should prevent this entire class of bugs and throw and error. But using the MOI wrapper is for experts, and they can use it with care. Let me find the discussion. I remember writing it all up somewhere...

**Comment by odow** [comment]
See https://github.com/jump-dev/JuMP.jl/issues/1739, https://github.com/jump-dev/JuMP.jl/pull/1766 but this isn't what I'm thinking of.

I think we decided that:

 * At the solver level, some solvers store solutions and allow querying them even if you have modified (some of) the model.
 * We don't want to enforce an `OptimizeNotCalled` error at the MOI level because experts may want the low-level solver behavior
 * At the JuMP level, accidentally querying old solutions or without calling `optimize!` seems like a _very_ common source of bugs, particularly for new users
 * To support users, JuMP added a readable error message

I regard changing the error message to the solver-default as "breaking" because it is likely to lead to confusion for users. And people may be relying on errors being thrown. Silently starting to return incorrect solutions (even if the user calls `value` without calling `optimize!` in error) seems like a bad direction to move in.
